### PR TITLE
Add support for OCaml compiler version variants

### PR DESCRIPTION
**Pull request by rivy**
- eg, "4.10.1+musl+flambda"
- supports darwin, unix, and windows
- some variable name refactoring included to improve readability

### CODE REVIEW

**Comment by rivy** [comment]
Apologies for the GHA build noise.
I didn't have actions enabled on my fork, so I ended up iterating on the PR.

I'm unsure of why the workflow job "build (ubuntu-18.04, 4.10.1+musl+flambda)" is failing. "build (ubuntu-20.04, 4.10.1+musl+flambda)" is working, and the error message about missing "musl-tools" is odd as I see it being installed earlier in the job. Ideas?

I'm pursuing this to enable an easier 32-bit build for Windows `unison`, but that job is having compile problems as well. I'll review and iterate on that one...

**Comment by rivy** [comment]
All working except "build (ubuntu-18.04, 4.10.1+musl+flambda)" which complains about missing "musl-tools" (which appears to be installed just above). Thoughts?

**Comment by rivy** [comment]
So, I found the problem with "build (ubuntu-18.04, 4.10.1+musl+flambda)". There is a static linking bug in the musl-tools package within Ubuntu 18.04 (see <https://github.com/ocaml/ocaml/issues/9131#issuecomment-599765888>). I've pushed a fix. 🤞

**Comment by rivy** [comment]
I boiled down and polished the PR to small set of explanatory commits. I'm happy to modify the PR based on any feedback.

I'm still working on getting a more complicated package ('lablgtk') installed (see <https://github.com/CICD-tools/setup-ocaml/tree/x.variants>). I'd like to get that working before I finalize this PR as the process may point to areas which require a bit more tweak.

**Comment by rivy** [comment]
ping @smorimoto @dra27 @avsm , this is ready for review! And, I'm happy to make any changes.

I've included a few tweaks for problems that I encountered while developing installation recipes for various packages, as well as fixes allowing for continued general `bash` scripting after Cygwin/OCaml is installed. I've also included a wide variety of platforms/variants and multiple packages (including the complicated 'lablgtk' package) for wide-ranging testing purposes. I expect that you would likely prefer a smaller subset of variants and packages to decrease the workflow build time.

Thanks for looking at this PR.
I'm going to now attempt to incorporate this into a set of auto-build recipes for [`unison`](https://github.com/bcpierce00/unison).

**Comment by smorimoto** [comment]
I didn't check this because I was a little busy with illness and work. I think we need to remove some unnecessary diff. And yaml for testing is too complicated and I'm worried about the cost of maintaining it. Also, it's very hard to read. Even though it could be just a matter of taste.

**Comment by smorimoto** [comment]
I will check the implementation later! Thanks anyway @rivy!

**Comment by rivy** [comment]
> I think we need to remove some unnecessary diff.

I'm happy to refactor any sections that trouble you.

> And yaml for testing is too complicated and I'm worried about the cost of maintaining it. Also, it's very hard to read. Even though it could be just a matter of taste.

The installation of  'lablgtk' was a requirement for `unison` and is included here only as a proof-of-concept for a more complicated package installation and as an example of how to use portable 'bash' shell scripts after the OCaml installation step. It can easily be removed. To be fair, some of the "readability" issues are because `yarn fmt` is word-wrapping some of the configuration variable lines.

I'm happy to rework the package installation tests to be whatever set you'd like to include. Notably, the 'yaml' package is not windows MSVC compatible, hence the necessary gating. A list of possible packages which do have at least some windows compatibility is on the [Ocaml for Windows ~ Packages](https://fdopen.github.io/opam-repository-mingw/packages) page. Any small set of suitably complex packages would seem to be reasonable.

I do hope that with this PR, more packages will start adding and testing windows compatibility.

**Comment by rivy** [comment]
I refactored the 'lablgtk' prerequisite setup and install steps into a single, more streamlined step (for all platforms).
It still might be too much and be better as a documentation example, but this is a better version either way.

File: .github/workflows/workflow.yml [hunk 1]
```diff
@@ -55,9 +35,97 @@ jobs:
...
+        uses: actions/checkout@v2
+
+      - name: Initialize workflow variables
+        id: vars
+        shell: bash
+        run: |
+          outputs() { for var in "$@" ; do echo steps.vars.outputs.${var}="${!var}" ; echo ::set-output name=${var}::${!var} ; done ; }
+          MINGW_ARCH='x86_64' ; case '${{ matrix.job.ocaml-version }}' in *+mingw32*) MINGW_ARCH='i686' ;; *+mingw64*) MINGW_ARCH='x86_64' ;; esac
+          MSVC_ARCH='' ; case '${{ matrix.job.ocaml-version }}' in *+msvc32*) MSVC_ARCH='x86' ;; *+msvc64*) MSVC_ARCH='x64' ;; esac
+          outputs MINGW_ARCH MSVC_ARCH
```
Comment (smorimoto) [inline .github/workflows/workflow.yml:82]:
```suggestion
          outputs() {
            for var in "$@"; do
              echo steps.vars.outputs."${var}"="${!var}"
              echo ::set-output name="${var}"::"${!var}"
            done
          }
          MINGW_ARCH='x86_64'
          case '${{ matrix.job.ocaml-version }}' in
            *+mingw32*) "$MINGW_ARCH"='i686' ;;
            *+mingw64*) "$MINGW_ARCH"='x86_64' ;;
          esac
          MSVC_ARCH=''
          case '${{ matrix.job.ocaml-version }}' in
            *+msvc32*) "$MSVC_ARCH"='x86' ;;
            *+msvc64*) "$MSVC_ARCH"='x64' ;;
          esac
          outputs "$MINGW_ARCH" "$MSVC_ARCH"
```

**Comment by smorimoto** [comment]
Yes, I think so. I re-triggered that.

File: src/install-ocaml-unix.sh [hunk 2]
```diff
@@ -5,8 +5,26 @@ set -ex
...
+OCAML_VV="$OCAML_VERSION"
+if [ "$OCAML_VARIANT" != "" ]; then OCAML_VV="$OCAML_VERSION+$OCAML_VARIANT" ; fi
 
-if [ "$CURRENT_OCAML" != "$1" ]; then
-  opam switch set "$1" || opam switch create "$1" "$1"
+CURRENT_OCAML=$(opam list ocaml-variants --installed --columns version --short --color=never)
+if [ -z "$CURRENT_OCAML" ]; then CURRENT_OCAML=$(opam info ocaml --field=version --color=never) ; fi
+
+OS_ID=$(eval echo $(cat /etc/os-release | grep -Po "(?<=^ID=)(.*)$"))
+OS_VERSION=$(eval echo $(cat /etc/os-release | grep -Po "(?<=^VERSION_ID=)(.*)$"))
```
Comment (smorimoto) [inline src/install-ocaml-unix.sh:17]:
```suggestion
OS_ID=$(eval echo "$(grep -Po "(?<=^ID=)(.*)$" </etc/os-release)")
OS_VERSION=$(eval echo "$(grep -Po "(?<=^VERSION_ID=)(.*)$" </etc/os-release)")
```

File: .github/workflows/workflow.yml (hunk 1 above)
Comment (rivy) [reply .github/workflows/workflow.yml:82]: Will do.

File: src/install-ocaml-unix.sh (hunk 2 above)
Comment (rivy) [reply src/install-ocaml-unix.sh:17]: Also, will do.

File: .github/workflows/workflow.yml (hunk 1 above)
Comment (rivy) [reply .github/workflows/workflow.yml:82]: The last line use of outputs is incorrect. `outputs MINGW_ARCH MSVC_ARCH` is the correct usage (variable names not values) and is left as is...

**Comment by rivy** [comment]
PR feedback integrated as "fixup!" commits which I'll squash down when this is ready for merge.
Also, rebased onto updated 'master' branch.

**Comment by smorimoto** [comment]
Sorry for the late review. (I have been pretty busy at work lately.) But I will review this weekend.

**Comment by rivy** [comment]
Ping @smorimoto , just checking in ... Any issues?
We're already using this via a fork for [`unison`](https://github.com/bcpierce00/unison).

**Comment by smorimoto** [comment]
I will check this today.

**Comment by smorimoto** [comment]
@rivy Can you enable the "Allow edits from maintainers" option?
https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork
I have resolved the conflict in my local and made some improvements, but I can't push.

**Comment by rivy** [comment]
@smorimoto , the option doesn't exit for this PR.
Let me know what location/branch/commits you would like pushed on top and I'll do it directly from your repo.

**Comment by smorimoto** [comment]
@rivy It must exist. Anyway, OK. Please synchronize with this branch, or I can open a new PR.

**Comment by rivy** [comment]
> @rivy It must exist.

It usually does... but no, it doesn't exist for this PR.
 


If it did, you'd be able to enable it as maintainer.

I've synchronized the commit tree (barring 84b1408) and fixed two problems.
I'm happy to help slim down the workflow, but 84b1408 rips out all of the 32-bit and windows variant testing.

**Comment by smorimoto** [comment]
Hmm... It's too weird...
I don't really think it's necessary to check all such patterns with CI. (That's because there is no end to it.)

**Comment by smorimoto** [comment]
FYI: I've implemented this in the actions-ml repository already. It will be released as a new version in the near future, so if you are thinking of doing some additional work, it may be duplicated work.

**Comment by smorimoto** [comment]
Variant will be supported in #66.

**Comment by rivy** [comment]
I hope that you will implement it successfully.

**Comment by smorimoto** [comment]
Well, it's already supported in the v2 branch.

**Pull request closed without merging**